clang/test/Driver/cl-cxx20-modules.cppm

```diff
@@ -0,0 +1,65 @@
+// REQUIRES: system-windows
+
+// RUN: rm -rf %t
+// RUN: mkdir -p %t
+// RUN: split-file %s %t
+
+// RUN: %clang_cl /std:c++20 --precompile "%/t/Hello.cppm" "/Fo%/t/Hello.pcm"
```

Comment:
This outputs `Hello.pcm` in `<build path>\tools\clang\test\Driver` instead of in the temporary directory specified by `%t` and the next `RUN` command fails. I'm not the most familiar with the LLVM testing infrastructure; I'd like this to be reviewed.

Reply:
We don't need to do that in the test of Drivers. We can pass `-###` here (You can look at other test cases in clang/test/Driver for an example) and check for the output. If `-###` is specified, it wouldn't run actually but only print the outputs.

Reply:
I've looked at the other tests but I don't quite follow how `-###` works—could you describe an example, please? Thanks!

Reply:
Maybe something like:

```
// RUN: %clang_cl /std:c++20 --precompile -### -- %s 2>&1 | FileCheck  %s
// CHECK: --precompile
```

and we can do such things for `-fmodule-file=` and `-fprebuilt-module-path=`

Reply:
So, I've tried to write the test like you've suggested exactly, but I end up with these errors:

```
  Testing: 
  FAIL: Clang :: Driver/cl-cxx20-modules.cppm (1 of 19831)
  ******************** TEST 'Clang :: Driver/cl-cxx20-modules.cppm' FAILED ********************
  Exit Code: 1
  
  Command Output (stdout):
  --
  # RUN: at line 1
  r:\llvm-project\build\x64-release\bin\clang.exe --driver-mode=cl /std:c++20 --precompile -### -- R:\llvm-project\clang\test\Driver\cl-cxx20-modules.cppm 2>&1 | r:\llvm-project\build\x64-release\bin\filecheck.exe  R:\llvm-project\clang\test\Driver\cl-cxx20-modules.cppm
  # executed command: 'r:\llvm-project\build\x64-release\bin\clang.exe' --driver-mode=cl /std:c++20 --precompile '-###' -- 'R:\llvm-project\clang\test\Driver\cl-cxx20-modules.cppm'
  # executed command: 'r:\llvm-project\build\x64-release\bin\filecheck.exe' 'R:\llvm-project\clang\test\Driver\cl-cxx20-modules.cppm'
  # .---command stderr------------
  # | // CHECK: --precompile
  # |           ^
  # | <stdin>:1:1: note: scanning from here
  # | clang version 19.0.0git (https://github.com/sharadhr/llvm-project.git e6386de2ef7dcb4650073c81e2da99bbd3446c99)
  # | ^
  # | <stdin>:6:744: note: possible intended match here
  # |  "D:\\Libraries\\Documents\\_Documents\\Repositories\\llvm-project\\build\\x64-release\\bin\\clang.exe" "-cc1" "-triple" "x86_64-pc-windows-msvc19.40.33812" "-emit-module-interface" "-disable-free" "-clear-ast-before-backend" "-disable-llvm-verifier" "-discard-value-names" "-main-file-name" "cl-cxx20-modules.cppm" "-mrelocation-model" "pic" "-pic-level" "2" "-mframe-pointer=none" "-relaxed-aliasing" "-fmath-errno" "-ffp-contract=on" "-fno-rounding-math" "-mconstructor-aliases" "-fms-volatile" "-funwind-tables=2" "-target-cpu" "x86-64" "-mllvm" "-x86-asm-syntax=intel" "-tune-cpu" "generic" "-D_MT" "-flto-visibility-public-std" "--dependent-lib=libcmt" "--dependent-lib=oldnames" "-stack-protector" "2" "-fdiagnostics-format" "msvc" "-fdebug-compilation-dir=R:\\llvm-project\\build\\x64-release\\tools\\clang\\test\\Driver" "-fcoverage-compilation-dir=R:\\llvm-project\\build\\x64-release\\tools\\clang\\test\\Driver" "-resource-dir" "D:\\Libraries\\Documents\\_Documents\\Repositories\\llvm-project\\build\\x64-release\\lib\\clang\\19" "-internal-isystem" "D:\\Libraries\\Documents\\_Documents\\Repositories\\llvm-project\\build\\x64-release\\lib\\clang\\19\\include" "-internal-isystem" "C:\\Program Files\\Microsoft Visual Studio\\2022\\Enterprise\\VC\\Tools\\MSVC\\14.40.33807\\include" "-internal-isystem" "C:\\Program Files\\Microsoft Visual Studio\\2022\\Enterprise\\VC\\Tools\\MSVC\\14.40.33807\\ATLMFC\\include" "-internal-isystem" "C:\\Program Files\\Microsoft Visual Studio\\2022\\Enterprise\\VC\\Auxiliary\\VS\\include" "-internal-isystem" "C:\\Program Files (x86)\\Windows Kits\\10\\include\\10.0.22621.0\\ucrt" "-internal-isystem" "C:\\Program Files (x86)\\Windows Kits\\10\\\\include\\10.0.22621.0\\\\um" "-internal-isystem" "C:\\Program Files (x86)\\Windows Kits\\10\\\\include\\10.0.22621.0\\\\shared" "-internal-isystem" "C:\\Program Files (x86)\\Windows Kits\\10\\\\include\\10.0.22621.0\\\\winrt" "-internal-isystem" "C:\\Program Files (x86)\\Windows Kits\\10\\\\include\\10.0.22621.0\\\\cppwinrt" "-internal-isystem" "C:\\Program Files (x86)\\Windows Kits\\NETFXSDK\\4.8\\include\\um" "-fdeprecated-macro" "-ferror-limit" "19" "-fno-use-cxa-atexit" "-fms-extensions" "-fms-compatibility" "-fms-compatibility-version=19.40.33812" "-std=c++20" "-fno-implicit-modules" "-fskip-odr-check-in-gmf" "-faddrsig" "-o" "cl-cxx20-modules.pcm" "-x" "c++" "R:\\llvm-project\\clang\\test\\Driver\\cl-cxx20-modules.cppm"
  # |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        ^
  # | 
  # | Input file: <stdin>
  # | Check file: R:\llvm-project\clang\test\Driver\cl-cxx20-modules.cppm
  # | 
  # | -dump-input=help explains the following input dump.
  # | 
  # | Input was:
  # | <<<<<<
  # |            1: clang version 19.0.0git (https://github.com/sharadhr/llvm-project.git e6386de2ef7dcb4650073c81e2da99bbd3446c99) 
  # |            2: Target: x86_64-pc-windows-msvc 
  # | check:2'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # |            3: Thread model: posix 
  # | check:2'0     ~~~~~~~~~~~~~~~~~~~~
  # |            4: InstalledDir: D:\Libraries\Documents\_Documents\Repositories\llvm-project\build\x64-release\bin 
  # | check:2'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # |            5:  (in-process) 
  # | check:2'0     ~~~~~~~~~~~~~~
  # |            6:  "D:\\Libraries\\Documents\\_Documents\\Repositories\\llvm-project\\build\\x64-release\\bin\\clang.exe" "-cc1" "-triple" "x86_64-pc-windows-msvc19.40.33812" "-emit-module-interface" "-disable-free" "-clear-ast-before-backend" "-disable-llvm-verifier" "-discard-value-names" "-main-file-name" "cl-cxx20-modules.cppm" "-mrelocation-model" "pic" "-pic-level" "2" "-mframe-pointer=none" "-relaxed-aliasing" "-fmath-errno" "-ffp-contract=on" "-fno-rounding-math" "-mconstructor-aliases" "-fms-volatile" "-funwind-tables=2" "-target-cpu" "x86-64" "-mllvm" "-x86-asm-syntax=intel" "-tune-cpu" "generic" "-D_MT" "-flto-visibility-public-std" "--dependent-lib=libcmt" "--dependent-lib=oldnames" "-stack-protector" "2" "-fdiagnostics-format" "msvc" "-fdebug-compilation-dir=R:\\llvm-project\\build\\x64-release\\tools\\clang\\test\\Driver" "-fcoverage-compilation-dir=R:\\llvm-project\\build\\x64-release\\tools\\clang\\test\\Driver" "-resource-dir" "D:\\Libraries\\Documents\\_Documents\\Repositories\\llvm-project\\build\\x64-release\\lib\\clang\\19" "-internal-isystem" "D:\\Libraries\\Documents\\_Documents\\Repositories\\llvm-project\\build\\x64-release\\lib\\clang\\19\\include" "-internal-isystem" "C:\\Program Files\\Microsoft Visual Studio\\2022\\Enterprise\\VC\\Tools\\MSVC\\14.40.33807\\include" "-internal-isystem" "C:\\Program Files\\Microsoft Visual Studio\\2022\\Enterprise\\VC\\Tools\\MSVC\\14.40.33807\\ATLMFC\\include" "-internal-isystem" "C:\\Program Files\\Microsoft Visual Studio\\2022\\Enterprise\\VC\\Auxiliary\\VS\\include" "-internal-isystem" "C:\\Program Files (x86)\\Windows Kits\\10\\include\\10.0.22621.0\\ucrt" "-internal-isystem" "C:\\Program Files (x86)\\Windows Kits\\10\\\\include\\10.0.22621.0\\\\um" "-internal-isystem" "C:\\Program Files (x86)\\Windows Kits\\10\\\\include\\10.0.22621.0\\\\shared" "-internal-isystem" "C:\\Program Files (x86)\\Windows Kits\\10\\\\include\\10.0.22621.0\\\\winrt" "-internal-isystem" "C:\\Program Files (x86)\\Windows Kits\\10\\\\include\\10.0.22621.0\\\\cppwinrt" "-internal-isystem" "C:\\Program Files (x86)\\Windows Kits\\NETFXSDK\\4.8\\include\\um" "-fdeprecated-macro" "-ferror-limit" "19" "-fno-use-cxa-atexit" "-fms-extensions" "-fms-compatibility" "-fms-compatibility-version=19.40.33812" "-std=c++20" "-fno-implicit-modules" "-fskip-odr-check-in-gmf" "-faddrsig" "-o" "cl-cxx20-modules.pcm" "-x" "c++" "R:\\llvm-project\\clang\\test\\Driver\\cl-cxx20-modules.cppm" 
  # | check:2'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # | check:2'1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            ?                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          possible intended match
  # | >>>>>>
  # `-----------------------------
R:\llvm-project\build\x64-release\EXEC : # error : command failed with exit status: 1
```

I'm not quite sure what to do here; any advice is welcome.

Reply:
I notice there's a very long test in `cl-options.c` that seem to test Clang-specific options. Could I just slide these three options in the long list of `//RUN: -fsomeoption`? In particular, this part: https://github.com/llvm/llvm-project/blob/d48d4805f792adbdac58d480f890449def4964ea/clang/test/Driver/cl-options.c#L724-L733

Reply:
Oh, sorry. It turns to be my bad. We should check `-emit-module-interface` for `--precompile` instead. I didn't think enough. And since we will check each line once with a `CHECK` but the output of `-###` is going to be in one line. So it may not be good to have 3 `CHECK` in one file. We can either split them by `split-file` or using different prefix for `FileCheck`.

Reply:
Ah. I found a nice example in `cl-pch.cpp` that does this. Hopefully this is what is required.